app/lib/analyzer/backend.dart

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:math';
+
+import 'package:gcloud/db.dart';
+
+import '../shared/utils.dart';
+
+import 'models.dart';
+
+final Random _random = new Random.secure();
+
+/// Datastore-related access methods for the analyzer service
+class AnalysisBackend {
+  final DatastoreDB db;
+
+  AnalysisBackend(this.db);
+
+  /// Gets the [Analysis] object from the Datastore. [version] and [analysis] is
+  /// optional, when they are missing the analysis from the latest rollup will
+  /// be returned.
+  Future<Analysis> getAnalysis(String package,
+      [String version, String analysis]) async {
+    final Key packageKey = db.emptyKey.append(PackageAnalysis, id: package);
+
+    Future<Analysis> loadRollup(AnalysisRollup rollup) async {
+      if (rollup == null) return null;
+      return (await db.lookup([rollup.analysisKey]))[0];
+    }
+
+    if (version == null) {
+      final list = await db.lookup([packageKey]);
+      return loadRollup(list[0] as PackageAnalysis);
+    }
+
+    // version was set
+    final Key versionKey =
+        packageKey.append(PackageVersionAnalysis, id: version);
+
+    if (analysis == null) {
+      final list = await db.lookup([versionKey]);
+      return loadRollup(list[0] as PackageVersionAnalysis);
+    }
+
+    // analysis was set
+    final Key analysisKey = versionKey.append(Analysis, id: analysis);
+    final List list = await db.lookup([analysisKey]);
+    return list[0];
+  }
+
+  /// Create a new [Analysis] object with the current timestamp.
+  Analysis prepareAnalysis(String package, String version) {
+    final DateTime now = new DateTime.now().toUtc();
+    final String id = _generateAnalysisId(now);
+    final Key packageKey = db.emptyKey.append(PackageAnalysis, id: package);
+    final Key versionKey =
+        packageKey.append(PackageVersionAnalysis, id: version);
+    return new Analysis()
+      ..parentKey = versionKey
+      ..id = id
+      ..packageKey = packageKey
+      ..versionKey = versionKey
```

Comment:
These two "back pointing" keys are not necessary at all, since the `Analysis.key` contains all the necessary information.

We might still want to store `Analysis.packageName` and `Analysis.versionName` as Strings though, for ease of querying `Analysis` objects and because it makes semantically.